File: typescipt/existing/src/specifications/specificationRunnerContext.ts






namespace Lexy.Compiler.Specifications;

public class SpecificationRunnerContext : ISpecificationRunnerContext, IDisposable
{
   private readonly List<ISpecificationFileRunner> fileRunners = new();

   private readonly ILogger<SpecificationRunnerContext> logger;

   public SpecificationRunnerContext(ILogger<SpecificationRunnerContext> logger)
   {
     this.logger = logger;
   }

   public void Dispose()
   {
     foreach (var fileRunner in fileRunners) fileRunner.Dispose();
   }

   //public IList<string> Messages { get; } = new List<string>();

   public int Failed { get; private set; }

   public IReadOnlyCollection<ISpecificationFileRunner> FileRunners => fileRunners;

   public void Fail(Scenario scenario, string message)
   {
     Failed++;

     var log = $"- FAILED - {scenario.Name}: {message}";

     Console.WriteLine(log);
     logger.LogError(log);
   }

   public void LogGlobal(string message)
   {
     Console.WriteLine(Environment.NewLine + message + Environment.NewLine);
     logger.LogInformation(message);
   }

   public void Log(string message)
   {
     var log = $" {message}";
     Console.WriteLine(log);
     logger.LogInformation(log);
   }

   public void Success(Scenario scenario)
   {
     var log = $"- SUCCESS - {scenario.Name}";
     Console.WriteLine(log);
     logger.LogInformation(log);
   }

   public void Add(ISpecificationFileRunner fileRunner)
   {
     fileRunners.Add(fileRunner);
   }

   public IEnumerable<IScenarioRunner> FailedScenariosRunners()
   {
     return fileRunners
       .SelectMany(runner => runner.ScenarioRunners)
       .Where(scenario => scenario.Failed);
   }

   public int CountScenarios()
   {
     return FileRunners.Select(fileRunner => fileRunner.CountScenarioRunners())
       .Aggregate((value, total) => value + total);
   }
}

File: src/Lexy.Compiler/Specifications/SpecificationRunnerContext.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Lexy.Compiler.Infrastructure;
using Lexy.Compiler.Language.Scenarios;
using Lexy.RunTime;
using Microsoft.Extensions.Logging;

namespace Lexy.Compiler.Specifications;

public class SpecificationRunnerContext : ISpecificationRunnerContext
{
    private readonly List<ISpecificationFileRunner> fileRunners = new();
    private readonly List<SpecificationsLogEntry> logEntries = new();
    private readonly ILogger<SpecificationsRunner> logger;
    private readonly DateTime startTimestamp;

    public int Failed { get; private set; }
    public IEnumerable<SpecificationsLogEntry> LogEntries => logEntries;
    public IReadOnlyCollection<ISpecificationFileRunner> FileRunners => fileRunners;

    public SpecificationRunnerContext(ILogger<SpecificationsRunner> logger)
    {
        startTimestamp = DateTime.Now;
        this.logger = logger;
    }

    public void Fail(Scenario scenario, string message, IEnumerable<string> errors)
    {
        Failed++;

        var entry = new SpecificationsLogEntry(scenario.Reference, scenario, true,
            $"FAILED - {scenario.Name}: {message}", errors);
        logEntries.Add(entry);

        logger.LogError("- FAILED  - {ScenarioName}: {Message}", scenario.Name, message);
        errors?.ForEach(message => this.logger.LogInformation("  {Message}", message));
    }

    public void LogGlobal(string message)
    {
        var entry = new SpecificationsLogEntry(null, null, false, message);
        logEntries.Add(entry);
        logger.LogInformation("{Message}", message);
    }

    public void LogTimeSpent()
    {
        var difference = DateTime.Now.Subtract(this.startTimestamp).TotalMilliseconds;
        var message = $"Time: {difference} milliseconds";

        var entry = new SpecificationsLogEntry(null, null, false, message);
        logEntries.Add(entry);
        logger.LogInformation("Time: {Difference} milliseconds", difference);
    }

    public void Success(Scenario scenario, IEnumerable<ExecutionLogEntry> logging = null)
    {
        var entry = new SpecificationsLogEntry(scenario.Reference, scenario, false, $"SUCCESS - {scenario.Name}", null,
            logging);
        logEntries.Add(entry);
        logger.LogInformation("- SUCCESS - {ScenarioName}", scenario.Name);
    }

    public void Add(ISpecificationFileRunner fileRunner)
    {
        fileRunners.Add(fileRunner);
    }

    public IEnumerable<IScenarioRunner> FailedScenariosRunners()
    {
        return fileRunners
            .SelectMany(runner => runner.ScenarioRunners)
            .Where(scenario => scenario.Failed);
    }

    public int CountScenarios()
    {
        return FileRunners.Select(fileRunner => fileRunner.CountScenarioRunners())
            .Aggregate((value, total) => value + total);
    }
}